ambry-rest/src/test/java/com.github.ambry.rest/NettyResponseChannelTest.java

```diff
@@ -310,6 +330,76 @@ public void setStatusTest() {
   }
 
+  /**
+   * Tests that HEAD returns no body in error responses.
+   */
+  @Test
+  public void noBodyForHeadTest() {
+    EmbeddedChannel channel = createEmbeddedChannel();
+    for (Map.Entry<RestServiceErrorCode, HttpResponseStatus> entry : REST_SERVICE_ERROR_CODE_TO_HTTP_RESPONSE_STATUS_MAP
+        .entrySet()) {
+      HttpHeaders httpHeaders = new DefaultHttpHeaders();
+      httpHeaders.set(MockNettyMessageProcessor.REST_SERVICE_ERROR_CODE_HEADER_NAME, entry.getKey());
+      channel.writeInbound(RestTestUtils
+          .createRequest(HttpMethod.HEAD, TestingUri.OnResponseCompleteWithRestException.toString(), httpHeaders));
+      HttpResponse response = (HttpResponse) channel.readOutbound();
+      assertEquals("Unexpected response status", entry.getValue(), response.getStatus());
+      if (response instanceof FullHttpResponse) {
+        // assert that there is no content
+        assertEquals("The response should not contain content", 0,
+            ((FullHttpResponse) response).content().readableBytes());
+      } else {
+        HttpContent content = (HttpContent) channel.readOutbound();
+        assertTrue("End marker should be received", content instanceof LastHttpContent);
+      }
+      assertNull("There should be no more data in the channel", channel.readOutbound());
+      boolean shouldBeAlive = !NettyResponseChannel.CLOSE_CONNECTION_ERROR_STATUSES.contains(entry.getValue());
+      assertEquals("Channel state (open/close) not as expected", shouldBeAlive, channel.isActive());
+      assertEquals("Connection header should be consistent with channel state", shouldBeAlive,
+          HttpHeaders.isKeepAlive(response));
+      if (!shouldBeAlive) {
+        channel = createEmbeddedChannel();
+      }
+    }
+    channel.close();
+  }
+
+  /**
+   * Tests keep-alive for different HTTP methods and error statuses.
+   */
+  @Test
+  public void keepAliveTest() {
+    HttpMethod[] HTTP_METHODS = {HttpMethod.POST, HttpMethod.GET, HttpMethod.HEAD, HttpMethod.DELETE};
+    EmbeddedChannel channel = createEmbeddedChannel();
+    for (HttpMethod httpMethod : HTTP_METHODS) {
+      for (Map.Entry<RestServiceErrorCode, HttpResponseStatus> entry : REST_SERVICE_ERROR_CODE_TO_HTTP_RESPONSE_STATUS_MAP
+          .entrySet()) {
+        HttpHeaders httpHeaders = new DefaultHttpHeaders();
+        httpHeaders.set(MockNettyMessageProcessor.REST_SERVICE_ERROR_CODE_HEADER_NAME, entry.getKey());
+        channel.writeInbound(RestTestUtils
+            .createRequest(httpMethod, TestingUri.OnResponseCompleteWithRestException.toString(), httpHeaders));
+        HttpResponse response = (HttpResponse) channel.readOutbound();
+        assertEquals("Unexpected response status", entry.getValue(), response.getStatus());
+        if (!(response instanceof FullHttpResponse)) {
+          // empty the channel
+          while (channel.readOutbound() != null) {
+            ;
```

Comment:
This is the formatting :( I always write this as
`while(condition);` 
and the formatter does this.